va/validation-authority.go

```diff
@@ -133,25 +143,55 @@ func (va ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdentif
 	}
 
 	httpRequest.Host = hostName
-	tr := &http.Transport{
-		// We are talking to a client that does not yet have a certificate,
-		// so we accept a temporary, invalid one.
-		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
-		// We don't expect to make multiple requests to a client, so close
-		// connection immediately.
-		DisableKeepAlives: true,
-	}
+	tr := *http.DefaultTransport.(*http.Transport)
+	// We are talking to a client that does not yet have a certificate,
+	// so we accept a temporary, invalid one.
+	tr.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	// We don't expect to make multiple requests to a client, so close
+	// connection immediately.
+	tr.DisableKeepAlives = true
+	// Intercept Dial in order to connect to the IP address we
+	// selected above.
+	defaultDial := tr.Dial
+	tr.Dial = func(_, _ string) (net.Conn, error) {
+		port := "80"
+		if va.TestMode {
+			port = "5001"
+		} else if scheme == "https" {
+			port = "443"
+		}
+		// Ignore the addr selected by net/http.
+		return defaultDial("tcp", net.JoinHostPort(addrs[0].String(), port))
+	}
+	var redirects []string
+	addrUsed := addrs[0]
```

Comment:
This is an insufficient approach to address selection.  At the very least, you should:
* Not connect to RFC 1918 addresses
* Reject IPv6 addresses
* Document your selection algorithm

I would be OK with having a config switch to allow IPv6, but we need to be able to turn it off, because our current network environment doesn't support it and we need to fail cleanly.

I would encourage you to study what the [default dialer does for address selection](https://go.googlesource.com/go/+/go1.4.2/src/net/dial.go#165) and document how your algorithm agrees or differs.